File: cpp/src/linked_list.cpp
#include "dlc_core/linked_list.h"

ListNode* reverseList(ListNode* head) {     // 206. Reverse Linked List
    ListNode* cur = nullptr;
    while (head) {
        cur = new ListNode(head->val, cur);
        head = head->next;
    }
    return cur;
}

ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {     // 21. Merge Two Sorted Lists
    ListNode* cur = new ListNode();
    ListNode* res = cur;

    while (list1 && list2) {
        if (list1->val < list2->val) {
            cur->next = new ListNode(list1->val);
            list1 = list1->next;
            cur = cur->next;
        } else {
            cur->next = new ListNode(list2->val);
            list2 = list2->next;
            cur = cur->next;
        }
    }
    cur->next = list1 ? list1 : list2;
    return res->next;
}

File: src/dlc/linked_list.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def hasCycle(head: Optional[ListNode]) -> bool:
    """141. Linked List Cycle"""
    fast = slow = head
    while fast and fast.next:
        fast = fast.next.next
        slow = slow.next
        if fast == slow:
            return True
    
    return False

def reverseList(head: Optional[ListNode]) -> Optional[ListNode]:
    """206. Reverse Linked List"""
    cur = None
    while head:
        cur = ListNode(head.val, cur)
        head = head.next
    return cur

def mergeTwoLists(list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    """21. Merge Two Sorted Lists"""

    dummy = cur = ListNode()

    while list1 and list2:
        if list1.val < list2.val:
            cur.next = ListNode(list1.val)
            cur = cur.next
            list1 = list1.next
        else:
            cur.next = ListNode(list2.val)
            cur = cur.next
            list2 = list2.next

    cur.next = list1 if list1 else list2
    return dummy.next
